### PR TITLE
feat: add GitHub action autodiscovery plugin

### PR DESCRIPTION
**Pull request by olblak**
This pullrequest add a new autodiscovery plugin for handling GitHub action update
It supports both the GitHub action and Gitea variant.

Here is an example of configuration

```
autodiscovery:
  scmid: default
  actionid:  default
  crawlers:
    github/action:
      credentials:
        "code.forgejo.com":
          kind: gitea
          token: yyyy
        "github.com":
          kind: github
          token: '{{ requiredEnv "GITHUB_TOKEN" }}'
      versionfilter:
        kind: semver
        pattern: minor
      only:
        - actions:
            "updatecli/updatecli-action": ""
```

Please note that credentials are fetched from the environment variable by default.

For 'github.com', the default value is set to first environment detected
  1. "UPDATECLI_GITHUB_TOKEN"
  2. "GITHUB_TOKEN"
 
For 'gitea.com' and 'codeberg.org', the default value is set to first environment detected
  1. "UPDATECLI_GITHUB_TOKEN"
  2. "GITEA_TOKEN"

- [x] https://github.com/updatecli/updatecli/issues/1963
- [x] https://github.com/updatecli/updatecli/pull/1966
- [x] https://github.com/updatecli/updatecli/pull/2850
- [x] https://github.com/updatecli/updatecli/pull/2848

## Test

To test this pull request, you can run the following commands:

```shell
go test ./pkg/plugins/autodiscovery/githubaction/
go build -o bin/updatecli .
./bin/updatecli diff --config e2e/updatecli.d/success.d/autodiscovery/github/action
```

### Tradeoff

/

### Potential improvement

Potential improvement would be to improve the gitearelease and githubrelease Updatecli plugin to work without credentials such as https://github.com/updatecli/updatecli/issues/1417.

### CODE REVIEW

**Comment by olblak** [comment]
Spotted an interesting issue when using on Gitea/Forgejo.
I tested on https://codeberg.org/olblak/Reusable-workflows and realize that the way we control credentials can't work

**Comment by olblak** [comment]
I think I spot a bug.
It fails if the filename contains empty space

```
ERROR: .forgejo/workflows/Build container image.yaml: the yaml file does not exist
```

The problem is probably after the go templating that remote single quotes

File: pkg/core/pipeline/autodiscovery/main.go [hunk 1]
```diff
@@ -35,6 +36,7 @@ var (
 			"dockercompose": dockercompose.Spec{},
 			"dockerfile":    dockerfile.Spec{},
 			"flux":          flux.Spec{},
+			"github/action": githubaction.Spec{},
```
[the remark this answers is not in the document]
Comment (olblak) [reply pkg/core/pipeline/autodiscovery/main.go:39]: Thanks for the suggestion, I am going to allow using "github/action" and "gitea/action"

Regarding credentials, I have mixed feelings.
The credentials to use depends on the content of the action file for example:

```
name: Go
on:
  push:
    branches:
      - master
  pull_request:
jobs:
  lint:
    name: lint
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: https://github.com/actions/checkout@v4
      - name: setup Go
        uses: https://mygitea_instance.com/actions/setup-go@v4
        with:
          go-version: "1.23"
      - name: run linters
        run: |
          make lint
```

This manifest works on Gitea/Forgejo
We would need two credentials, one to interact with "https://github.com" and another one for "https://mygitea_instance.com"

But on GitHub, as far as I know, it's not possible to specify custom URL. That being said, I would expect GitHub enterprise customers to use their own action.

[the remark this answers is not in the document]
Comment (loispostula) [reply pkg/core/pipeline/autodiscovery/main.go:39]: I wasn't sure for the credentials, it makes a lot of sense!

**Comment by olblak** [comment]
I think the pullrequest is now ready.

**Review by loispostula — APPROVED** [review]

**Pull request merged**